### PR TITLE
Upgrade `k8s.io/*` to `v0.28`, `sigs.k8s.io/controller-runtime` to `v0.16`

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area open-source
/kind enhancement

**What this PR does / why we need it**:
This PR vendor k8s.io/*@v0.28.2, sigs.k8s.io/controller-runtime@v0.16.2 and sigs.k8s.io/controller-tools@v0.13.0 and adapts to the breaking / most notable changes

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/8382

**Special notes for your reviewer**:
In draft for -
- <strike> patch release of controller-runtime </strike>
-  <strike>depending on https://github.com/gardener/gardener/pull/8447 </strike>

**Release note**:

```noteworthy operator
The following golang dependencies have been upgraded, please consult the upstream release notes and [this issue](https://github.com/gardener/gardener/issues/8382) for guidance on upgrading your golang dependencies when vendoring this gardener version:
- `k8s.io/*` to `v0.28.2`
- `sigs.k8s.io/controller-runtime` to `v0.16.2`
- `sigs.k8s.io/controller-tools` to `v0.13.0`
```
```breaking developer
`leader-election-resource-lock` flag is dropped and the leader-election resource-lock is hard coded to leases.
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -61,7 +61,7 @@ require (
 	k8s.io/metrics v0.28.1
 	k8s.io/pod-security-admission v0.28.1
 	k8s.io/utils v0.0.0-20230406110748-d93618cff8a2
-	sigs.k8s.io/controller-runtime v0.15.2
+	sigs.k8s.io/controller-runtime v0.16.1
 	sigs.k8s.io/controller-runtime/tools/setup-envtest v0.0.0-20221212190805-d4f1e822ca11 // v0.14.1
```
Comment (rfranzke) [inline go.mod]: Should we also update this?

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 2]
```diff
@@ -257,9 +257,18 @@ func (o *Options) config(kubeAPIServerConfig *rest.Config, kubeClient *kubernete
 		}, nil
 	}
 
+	gardenerKubeClient, err := kubernetes.NewForConfig(gardenerAPIServerConfig.ClientConfig)
```
Comment (rfranzke) [inline cmd/gardener-apiserver/app/gardener_apiserver.go:260]: Why "gardenerKubeClient"? Isn't it just a "kubernetesClient", not specific to Gardener resources at all?

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 3]
```diff
@@ -257,9 +257,18 @@ func (o *Options) config(kubeAPIServerConfig *rest.Config, kubeClient *kubernete
...
+	if err != nil {
+		return nil, err
+	}
+	gardenerDynamicClient, err := dynamic.NewForConfig(gardenerAPIServerConfig.ClientConfig)
```
Comment (rfranzke) [inline cmd/gardener-apiserver/app/gardener_apiserver.go:264]: Just "dynamicClient"?

File: cmd/gardener-apiserver/app/gardener_apiserver.go [hunk 4]
```diff
@@ -433,9 +442,5 @@ func (o *Options) ApplyTo(config *apiserver.Config) error {
 		),
 	}
 
-	if err := o.Recommended.Etcd.Complete(config.GenericConfig.StorageObjectCountTracker, config.GenericConfig.DrainedNotify(), config.GenericConfig.AddPostStartHook); err != nil {
-		return err
-	}
```
Comment (rfranzke) [inline cmd/gardener-apiserver/app/gardener_apiserver.go:443]: Where is this gone/how is this replaced? 👀

File: vendor/sigs.k8s.io/controller-runtime/pkg/cache/cache.go [hunk 5]
```diff
@@ -427,7 +427,7 @@ func defaultOpts(config *rest.Config, opts Options) (Options, error) {
 		byObject.Transform = defaultedConfig.Transform
 		byObject.UnsafeDisableDeepCopy = defaultedConfig.UnsafeDisableDeepCopy
 
-		if byObject.Namespaces == nil {
+		if isNamespaced && byObject.Namespaces == nil {
```
Comment (rfranzke) [inline vendor/sigs.k8s.io/controller-runtime/pkg/cache/cache.go:430]: ? :)

File: cmd/gardener-apiserver/app/gardener_apiserver.go (hunk 4 above)
Comment (acumino) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:443]: It is now done inside `ApplyWithStorageFactoryTo`. ref - https://github.com/kubernetes/kubernetes/pull/118416

File: cmd/gardener-apiserver/app/gardener_apiserver.go (hunk 2 above)
Comment (acumino) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:260]: Here client is created from `gardenerAPIServerConfig.ClientConfig` that's why I named it `gardenerKubeClient`.
`gardenerAPIServerConfig ` has api codec for gardener APIs

File: cmd/gardener-apiserver/app/gardener_apiserver.go (hunk 3 above)
Comment (acumino) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:264]: Here client is created from `gardenerAPIServerConfig.ClientConfig` that's why I named it `gardenerDynamicClient `.

File: cmd/gardener-apiserver/app/gardener_apiserver.go (hunk 4 above)
Comment (rfranzke) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:443]: Thanks - do we need something like  https://github.com/kubernetes/kubernetes/pull/118416/files#diff-170c324cc9e649c56eba833146296ea83b4fe433cb1c05b63ca18578609da4d2R408 then as well?

**Comment by shafeeqes** [comment]
Please also update https://github.com/gardener/gardener/blob/00365d6c0fa6fb9053dec6f1e2fa489c53ef53cc/hack/prepare-envtest.sh#L21
and the version in integration tests like https://github.com/gardener/gardener/blob/638a8a8b4bcdd6419dac86b6867aeaedabf45bab/test/integration/resourcemanager/garbagecollector/garbagecollector_suite_test.go#L126

**Comment by shafeeqes** [comment]
`v0.16.2` also contains https://github.com/kubernetes-sigs/controller-runtime/pull/2472, I have not checked in detail, but from https://github.com/kubernetes-sigs/controller-runtime/pull/2425#issuecomment-1715710831, it seems like it fixes what https://github.com/kubernetes-sigs/controller-runtime/pull/2425 was trying to.

**Comment by acumino** [comment]
> Please also update
> 
> https://github.com/gardener/gardener/blob/00365d6c0fa6fb9053dec6f1e2fa489c53ef53cc/hack/prepare-envtest.sh#L21
> 
> 
> and the version in integration tests like
> https://github.com/gardener/gardener/blob/638a8a8b4bcdd6419dac86b6867aeaedabf45bab/test/integration/resourcemanager/garbagecollector/garbagecollector_suite_test.go#L126

It can only be done once 1.28 is support is available.

> v0.16.2 also contains https://github.com/kubernetes-sigs/controller-runtime/pull/2472, I have not checked in detail, but from https://github.com/kubernetes-sigs/controller-runtime/pull/2425#issuecomment-1715710831, it seems like it fixes what https://github.com/kubernetes-sigs/controller-runtime/pull/2425 was trying to.

I will check for it separately(separate PR).

File: cmd/gardener-apiserver/app/gardener_apiserver.go (hunk 4 above)
Comment (acumino) [reply cmd/gardener-apiserver/app/gardener_apiserver.go:443]: No, this is already passed in `ApplyWithStorageFactoryTo` second argument as part of `gardenerAPIServerConfig.Config`

File: pkg/operation/botanist/kubeapiserver_test.go [hunk 6]
```diff
@@ -640,17 +639,19 @@ var _ = Describe("KubeAPIServer", func() {
 		})
 	})
 
-	Describe("#ScaleKubeAPIServerToOne", func() {
-		deployment := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: "kube-apiserver", Namespace: seedNamespace}}
+	// TODO(acumino): Enable this test again when fakeclient start supporting scale subresource.
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver_test.go]: Is there an upstream issue or PR that we could link here?

Comment (acumino) [reply pkg/operation/botanist/kubeapiserver_test.go]: I have linked the issue in TODO. PTAL

**Review by rfranzke — COMMENTED** [review]
Thank you!

/lgtm
/approve

**Comment by acumino** [comment]
/hold
Something wrong is going on with GitHub changes pushed to the branch is not getting reflected in PR.

**Pull request merged**